juju/juju
# K8s AdmissionController for add missing Juju labels to objects created outside of Juju's k8s broker.

----
## Description of change

- Commit adds two new workers to Juju.

- First worker is responsible for running a Kubernetes controller that
  actively watches services accounts with applications labels in the
  models namespace. These service accounts are transformed into a map to
  be quered later by the second worker.

- The second worker extends the Juju http api-server by adding a new
  endpoint for each caas model that will process Kubernetes admission
  requests. Requests are sent from the Kubernetes api to our endpoint
  where we check if the user creating/modifying the resource is a Juju
  user. If so we make sure the object is annotated with the correct
  application it belongs to.

- This PR also adds some technical debt to fix where by it will only
  work properly on microk8s clusters. Other clusters will not work but
  won't add a noticable error in Juju or Kubernetes.

## QA steps

- Bootstrap microk8s cluster with juju (this is the only supported option at the moment)
- Install a k8s charm that creates a service account or install a new service account into the same namespace as the model with juju-app label set.
- Create a new kubernetes object with the above service account in the same namespace as the model.
- The resulting object should contain juju-app labels

## Documentation changes
To be done.
